Build failure in cross build to arm6l linux

**Describe the bug**

Build problem, link failure in `nix-store`

`nix build .#nix-ng-armv6l-unknown-linux-gnueabihf`

```
       > armv6l-unknown-linux-gnueabihf-g++  -o libnixstore.so libnixstore.so.p/binary-cache-store.cc.o libnixstore.so.p/build-result.cc.o libnixstore.so.p/build_derivation-goal.cc.o libnixstore.so.p/build_drv-output-substitution-goal.cc.o libnixstore.so.p/build_entry-points.cc.o libnixstore.so.p/build_goal.cc.o libnixstore.so.p/build_substitution-goal.cc.o libnixstore.so.p/build_worker.cc.o libnixstore.so.p/builtins_buildenv.cc.o libnixstore.so.p/builtins_fetchurl.cc.o libnixstore.so.p/builtins_unpack-channel.cc.o libnixstore.so.p/common-protocol.cc.o libnixstore.so.p/common-ssh-store-config.cc.o libnixstore.so.p/content-address.cc.o libnixstore.so.p/daemon.cc.o libnixstore.so.p/derivations.cc.o libnixstore.so.p/derived-path-map.cc.o libnixstore.so.p/derived-path.cc.o libnixstore.so.p/downstream-placeholder.cc.o libnixstore.so.p/dummy-store.cc.o libnixstore.so.p/export-import.cc.o libnixstore.so.p/filetransfer.cc.o libnixstore.so.p/gc.cc.o libnixstore.so.p/globals.cc.o libnixstore.so.p/http-binary-cache-store.cc.o libnixstore.so.p/indirect-root-store.cc.o libnixstore.so.p/keys.cc.o libnixstore.so.p/legacy-ssh-store.cc.o libnixstore.so.p/local-binary-cache-store.cc.o libnixstore.so.p/local-fs-store.cc.o libnixstore.so.p/local-overlay-store.cc.o libnixstore.so.p/local-store.cc.o libnixstore.so.p/log-store.cc.o libnixstore.so.p/machines.cc.o libnixstore.so.p/make-content-addressed.cc.o libnixstore.so.p/misc.cc.o libnixstore.so.p/names.cc.o libnixstore.so.p/nar-accessor.cc.o libnixstore.so.p/nar-info-disk-cache.cc.o libnixstore.so.p/nar-info.cc.o libnixstore.so.p/optimise-store.cc.o libnixstore.so.p/outputs-spec.cc.o libnixstore.so.p/parsed-derivations.cc.o libnixstore.so.p/path-info.cc.o libnixstore.so.p/path-references.cc.o libnixstore.so.p/path-with-outputs.cc.o libnixstore.so.p/path.cc.o libnixstore.so.p/pathlocks.cc.o libnixstore.so.p/posix-fs-canonicalise.cc.o libnixstore.so.p/profiles.cc.o libnixstore.so.p/realisation.cc.o libnixstore.so.p/remote-fs-accessor.cc.o libnixstore.so.p/remote-store.cc.o libnixstore.so.p/s3-binary-cache-store.cc.o libnixstore.so.p/serve-protocol-connection.cc.o libnixstore.so.p/serve-protocol.cc.o libnixstore.so.p/sqlite.cc.o libnixstore.so.p/ssh-store.cc.o libnixstore.so.p/ssh.cc.o libnixstore.so.p/store-api.cc.o libnixstore.so.p/store-reference.cc.o libnixstore.so.p/uds-remote-store.cc.o libnixstore.so.p/worker-protocol-connection.cc.o libnixstore.so.p/worker-protocol.cc.o libnixstore.so.p/linux_personality.cc.o libnixstore.so.p/unix_build_child.cc.o libnixstore.so.p/unix_build_hook-instance.cc.o libnixstore.so.p/unix_build_local-derivation-goal.cc.o libnixstore.so.p/unix_pathlocks.cc.o libnixstore.so.p/unix_user-lock.cc.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,libnixstore.so -fuse-ld=gold -Wl,--start-group /nix/store/3xc6gmhgllslbw2lhpj50ppysrm3lz1q-nix-util-armv6l-unknown-linux-gnueabihf-2.25.0pre/lib/libnixutil.so /nix/store/gan02jfaki8k9fzjbyzsdvvnlwy8kq1w-libarchive-armv6l-unknown-linux-gnueabihf-3.7.4-lib/lib/libarchive.so -pthread /nix/store/2a6ccf6r92va8a2xygx04n80228x1d8x-curl-armv6l-unknown-linux-gnueabihf-8.7.1/lib/libcurl.so /nix/store/2xhcvh6lvyr1n4251x7zf9p1grvbdj0r-libseccomp-armv6l-unknown-linux-gnueabihf-2.5.5-lib/lib/libseccomp.so /nix/store/xzn4i9mcpx6680z1c5vrslckgcpn70s2-sqlite-armv6l-unknown-linux-gnueabihf-3.45.3/lib/libsqlite3.so /nix/store/nj5mgi7vv512375q9kzmx8dycs8n1k2c-boost-armv6l-unknown-linux-gnueabihf-1.81.0/lib/libboost_container.so.1.81.0 -Wl,--end-group
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:635: error: undefined reference to '__atomic_fetch_add_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:418: error: undefined reference to '__atomic_fetch_add_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:635: error: undefined reference to '__atomic_fetch_add_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:635: error: undefined reference to '__atomic_fetch_add_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:481: error: undefined reference to '__atomic_store_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:505: error: undefined reference to '__atomic_load_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:505: error: undefined reference to '__atomic_load_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:505: error: undefined reference to '__atomic_load_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:426: error: undefined reference to '__atomic_fetch_sub_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:426: error: undefined reference to '__atomic_fetch_sub_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:426: error: undefined reference to '__atomic_fetch_sub_8'
```

**Additional context**

- Post-meson regression of https://github.com/NixOS/nix/issues/3113

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

--- a/flake.nix
+++ b/flake.nix
@@ -124,18 +124,36 @@
           # without "polluting" the top level "`pkgs`" attrset.
           # This also has the benefit of providing us with a distinct set of packages
           # we can iterate over.
-          nixComponents = lib.makeScope final.nixDependencies.newScope (import ./packaging/components.nix {
-            inherit (final) lib;
-            inherit officialRelease;
-            src = self;
-          });
+          nixComponents =
+            lib.makeScopeWithSplicing'
+              {
+                inherit (final) splicePackages;
+                inherit (final.nixDependencies) newScope;
+              }
+              {
+                otherSplices = final.generateSplicesForMkScope "nixComponents";
+                f = import ./packaging/components.nix {
+                  inherit (final) lib;
+                  inherit officialRelease;
+                  src = self;
+                };
+              };
 
           # The dependencies are in their own scope, so that they don't have to be
           # in Nixpkgs top level `pkgs` or `nixComponents`.
-          nixDependencies = lib.makeScope final.newScope (import ./packaging/dependencies.nix {
-            inherit inputs stdenv;
-            pkgs = final;
-          });
+          nixDependencies =
+            lib.makeScopeWithSplicing'
+              {
+                inherit (final) splicePackages;
+                inherit (final) newScope; # layered directly on pkgs, unlike nixComponents above
+              }
+              {
+                otherSplices = final.generateSplicesForMkScope "nixDependencies";
+                f = import ./packaging/dependencies.nix {
+                  inherit inputs stdenv;
+                  pkgs = final;
+                };
+              };
 
           nix = final.nixComponents.nix-cli;
 

